tools/decdk/README.md

```diff
@@ -0,0 +1,256 @@
+
+### Data Interfaces ("Props")
+
+jsii has a concept of "data interfaces", which are basically interfaces that do not have methods. For example, all construct "props" are data interfaces.
+
+> In some languages (Python, Ruby), if a method accepts a data interface as the last argument, interface properties can be used as keyword arguments in the method call. Other languages have a different idiomatic representation of data such as Java PoJos and Builders.
+
+deCDK maps data interfaces to closed JSON objects (no additional properties), and will recursively deconstruct all property types.
+
+### Primitives
+
+Strings, numbers, booleans, dates, lists and maps are all deconstructed 1:1 to their JSON representation.
+
+### Enums
+
+Enums are mapped to JSON schema enums.
+
+### References
+
+If deCDK encounters a reference to another __construct__ (a type that extends `cdk.Construct` or an interface that extends `cdk.IConstruct`), it will allow referencing it via a “Ref” intrinsic. For example, here's a definition of an ECS cluster that references a VPC:
+
+```yaml
+Resources:
+  VPC:
+    Type: "@aws-cdk/aws-ec2.VpcNetwork"
+    Properties:
+      maxAZs: 1
+  Cluster:
+    Type: "@aws-cdk/aws-ecs.Cluster"
+    Properties:
+      vpc:
+        Ref: VPC
+```
+
+### Enum-like Classes
+
+Based on the AWS Construct Library's consistent guidelines and conventions, which are also enforced by a tool we use called “awslint”, deCDK is also capable of expressive more complex idioms. For example, enum-like classes, which are classes that expose a set of static properties or methods can be mapped to JSON enums or method invocations. For example, this is how you define an AWS Lambda function in the CDK (TypeScript):
+
+```ts
+new lambda.Function(this, 'MyHandler', {
+  handler: 'index.handler',
+  runtime: lambda.Runtime.NodeJS810,
+  code: lambda.Code.asset('./src')
+});
+```
+
+And here's the deCDK version:
+
+```yaml
+MyHandler:
+  Type: "@aws-cdk/aws-lambda.Function"
+  Properties:
+    handler: index.handler
+    runtime: NodeJS810
+    code:
+      asset:
+        path: "./src"
+```
+
+### Polymorphism
+
+Due to the decoupled nature of AWS, The AWS Construct Library highly utilizes polymorphism to expose rich APIs to users. In many cases, APIs would accept an interface of some kind, and various AWS services provide an implementation for that interface. deCDK is able to find all concrete implementation of an interface or an abstract class and offer the user an enum-like experience. The following example shows how this approach can be used to define AWS Lambda events:
+
+```yaml
+MyHandler:
+  Type: "@aws-cdk/aws-lambda.Function"
+  Properties:
+    handler: index.handler
+    runtime: NodeJS810
+    code:
+      asset:
+        path: "./src"
+    events:
+    - DynamoEventSource:
```

Comment:
It's megamagic to me that this class comes from a different package. What if two packages share the same class name?

Reply:
You are right. This is actually not an accurate snippet, see [this](https://github.com/awslabs/aws-cdk/blob/fb431e28ea238badb8d256cd6a095e5a4b185c0b/tools/decdk/examples/lambda-events.json#L29)

Reply:
Updated README